module/zfs/zfs_vnops.c

```diff
@@ -3563,6 +3652,17 @@ zfs_rename(struct inode *sdip, char *snm, struct inode *tdip, char *tnm,
 	}
 
 	/*
+	 * If we are using project inheritance, we only allow renames
+	 * into our tree when the project IDs are the same; otherwise
+	 * tree quota mechanism would be circumvented.
```

Comment:
The comment implies that we don't just "hope" all the children will have the same project, we enforce that they do.  Is that a universal requirement - if ZFS_PROJINHERIT is set on a directory, all its descendants must have the same project?